packages/eservice-template-updater/src/eserviceTemplateUpdaterConsumerServiceV2.ts

```diff
@@ -0,0 +1,419 @@
+            }
+          );
+        }
+      );
+    })
+    .with(
+      { type: "EServiceTemplateEServiceDescriptionUpdated" },
+      async (msg) => {
+        const newDescription = getTemplateFromEvent(msg).eserviceDescription;
+
+        await commitUpdateToInstances(
+          msg,
+          refreshableToken,
+          correlationId,
+          async (instance, headers) => {
+            await catalogProcess.client.internalUpdateEServiceDescription(
+              { description: newDescription },
+              {
+                params: {
+                  eServiceId: instance.id,
+                },
+                headers,
+              }
+            );
+          }
+        );
+      }
+    )
+    .with({ type: "EServiceTemplateVersionAttributesUpdated" }, async (msg) => {
+      const eserviceTemplateVersion = getTemplateVersionFromEvent(msg);
+
+      const attributes = fromEServiceTemplateVersionV2(
+        eserviceTemplateVersion
+      ).attributes;
+
+      await commitUpdateToInstanceDescriptors(
+        msg,
+        refreshableToken,
+        correlationId,
+        async (instance, descriptor, headers) => {
+          await catalogProcess.client.internalUpdateDescriptorAttributes(
+            attributes,
+            {
+              params: {
+                eServiceId: instance.id,
+                descriptorId: descriptor.id,
+              },
+              headers,
+            }
+          );
+        }
+      );
+    })
+    .with({ type: "EServiceTemplateVersionQuotasUpdated" }, async (msg) => {
+      const eserviceTemplateVersion = getTemplateVersionFromEvent(msg);
+
+      await commitUpdateToInstanceDescriptors(
+        msg,
+        refreshableToken,
+        correlationId,
+        async (instance, descriptor, headers) => {
+          await catalogProcess.client.internalUpdateDescriptorVoucherLifespan(
+            { voucherLifespan: eserviceTemplateVersion.voucherLifespan },
+            {
+              params: {
+                eServiceId: instance.id,
+                descriptorId: descriptor.id,
+              },
+              headers,
+            }
+          );
+        }
+      );
+    })
+    .with({ type: "EServiceTemplateVersionDocumentAdded" }, async (msg) => {
+      const eserviceTemplateVersion = getTemplateVersionFromEvent(msg);
+
+      const docToAddToInstances = eserviceTemplateVersion.docs.find(
+        (d) => d.id === msg.data.documentId
+      );
+
+      if (!docToAddToInstances) {
+        return;
+      }
+
+      await commitUpdateToInstanceDescriptors(
+        msg,
+        refreshableToken,
+        correlationId,
+        async (instance, descriptor, headers) => {
+          const alreadyHasDoc = descriptor?.docs.some(
+            (d) => d.checksum === docToAddToInstances.checksum
+          );
+
+          if (alreadyHasDoc) {
+            return;
+          }
+
+          if (descriptor.state === "ARCHIVED") {
+            return;
+          }
```

Comment:
generally I prefer checks to be in one place only (in this case the catalog-process), but since this consumer may handle a large amount of eservices it is ok IMO to apply a preliminary filter on the document existence.
Anyway I recommend to keep all other checks in the process. For example the one on the descriptor state is only here

Reply:
I also returned early to avoid cloning the document without needing to do so